syncapi/storage/postgres/relations_table.go

```diff
@@ -0,0 +1,199 @@
+	selectMaxRelationIDStmt              *sql.Stmt
+}
+
+func NewPostgresRelationsTable(db *sql.DB) (tables.Relations, error) {
+	s := &relationsStatements{}
+	_, err := db.Exec(relationsSchema)
+	if err != nil {
+		return nil, err
+	}
+	if s.insertRelationStmt, err = db.Prepare(insertRelationSQL); err != nil {
+		return nil, err
+	}
+	if s.selectRelationsInRangeAscStmt, err = db.Prepare(selectRelationsInRangeAscSQL); err != nil {
+		return nil, err
+	}
+	if s.selectRelationsInRangeDescStmt, err = db.Prepare(selectRelationsInRangeDescSQL); err != nil {
+		return nil, err
+	}
+	if s.selectRelationsByTypeInRangeAscStmt, err = db.Prepare(selectRelationsByTypeInRangeAscSQL); err != nil {
+		return nil, err
+	}
+	if s.selectRelationsByTypeInRangeDescStmt, err = db.Prepare(selectRelationsByTypeInRangeDescSQL); err != nil {
+		return nil, err
+	}
+	if s.deleteRelationStmt, err = db.Prepare(deleteRelationSQL); err != nil {
+		return nil, err
+	}
+	if s.selectMaxRelationIDStmt, err = db.Prepare(selectMaxRelationIDSQL); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (s *relationsStatements) InsertRelation(
+	ctx context.Context, txn *sql.Tx, roomID, eventID, childEventID, relType string,
+) (streamPos types.StreamPosition, err error) {
+	err = sqlutil.TxStmt(txn, s.insertRelationStmt).QueryRowContext(
+		ctx, roomID, eventID, childEventID, relType,
+	).Scan(&streamPos)
+	return
+}
+
+func (s *relationsStatements) DeleteRelation(
+	ctx context.Context, txn *sql.Tx, roomID, childEventID string,
+) error {
+	stmt := sqlutil.TxStmt(txn, s.deleteRelationStmt)
+	_, err := stmt.ExecContext(
+		ctx, roomID, childEventID,
+	)
+	if err == sql.ErrNoRows {
+		return nil
+	}
+	return err
+}
+
+// SelectRelationsInRange returns a map rel_type -> []child_event_id
+func (s *relationsStatements) SelectRelationsInRange(
+	ctx context.Context, txn *sql.Tx, roomID, eventID, relType string,
+	r types.Range, limit int,
+) (map[string][]types.RelationEntry, types.StreamPosition, error) {
+	var lastPos types.StreamPosition
+	var stmt *sql.Stmt
+	var rows *sql.Rows
+	var err error
+	if relType != "" {
+		if r.Backwards {
+			stmt = sqlutil.TxStmt(txn, s.selectRelationsByTypeInRangeDescStmt)
+		} else {
+			stmt = sqlutil.TxStmt(txn, s.selectRelationsByTypeInRangeAscStmt)
+		}
+		rows, err = stmt.QueryContext(ctx, roomID, eventID, relType, r.Low(), r.High(), limit)
+	} else {
+		if r.Backwards {
+			stmt = sqlutil.TxStmt(txn, s.selectRelationsInRangeDescStmt)
+		} else {
+			stmt = sqlutil.TxStmt(txn, s.selectRelationsInRangeAscStmt)
+		}
+		rows, err = stmt.QueryContext(ctx, roomID, eventID, r.Low(), r.High(), limit)
+	}
+	if err != nil {
+		return nil, lastPos, err
+	}
+	defer internal.CloseAndLogIfError(ctx, rows, "selectRelationsInRange: rows.close() failed")
+	result := map[string][]types.RelationEntry{}
+	for rows.Next() {
+		var (
+			id           types.StreamPosition
+			roomID       string
+			childEventID string
+			relType      string
+		)
```

Comment:
I think we can move the variable declaration outside the loop, to avoid allocations.